src/lib/comms/sol-oic-client.c

```diff
@@ -244,12 +244,44 @@ _parse_platform_info_payload(struct sol_oic_platform_information *info,
 }
 
 static bool
+extract_device_id(CborValue *map, struct sol_buffer *device_id)
+{
+    CborValue value;
+    struct sol_str_slice slice;
+    int r;
+
+    sol_buffer_init_flags(device_id, NULL, 0, SOL_BUFFER_FLAGS_NO_NUL_BYTE);
+    if (cbor_value_map_find_value(map, SOL_OIC_KEY_DEVICE_ID, &value) != CborNoError)
+        return false;
+
+    if (cbor_value_is_byte_string(&value))
+        return cbor_value_dup_byte_string(&value, (uint8_t **)&device_id->data,
+            &device_id->used, NULL) == CborNoError;
+
+    if (cbor_value_is_text_string(&value)) {
+        r = cbor_value_dup_text_string(&value, (char **)&slice.data,
+            &slice.len, NULL) == CborNoError;
+        SOL_INT_CHECK_GOTO(r, < 0, error);
```

Comment:
error handling was wrong. Fixing that. 
And if cbor_value_dup_text_string fails, slice data is freed by function, so we don't need to free that.